keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/NUTS.java

```diff
@@ -0,0 +1,546 @@
+
+        return new NetworkSamples(samples, sampleCount);
+    }
+
+    private static BuiltTree BuildTree(List<Vertex<Double>> latentVertices,
+                                       final Map<String, Long> latentSetAndCascadeCache,
+                                       List<Vertex<?>> probabilisticVertices,
+                                       final List<? extends Vertex<?>> sampleFromVertices,
+                                       Map<String, Double> position,
+                                       Map<String, Double> gradient,
+                                       Map<String, Double> momentum,
+                                       double u,
+                                       int v,
+                                       int j,
+                                       double epsilon,
+                                       Random random
+    ) {
+
+        if (j == 0) {
+            //Base case—take one leapfrog step in the direction v
+
+            LeapFrogged leapfrog = leapfrog(
+                    latentVertices,
+                    latentSetAndCascadeCache,
+                    probabilisticVertices,
+                    position,
+                    gradient,
+                    momentum,
+                    epsilon * v
+            );
+
+            final double logOfMasterPAfterLeapfrog = getLogProb(probabilisticVertices);
+
+            final double logMpMinusMomentum = logOfMasterPAfterLeapfrog - 0.5 * dotProduct(leapfrog.momentum);
+            final int nPrime = u <= Math.exp(logMpMinusMomentum) ? 1 : 0;
+            final int sPrime = u < Math.exp(DELTA_MAX + logMpMinusMomentum) ? 1 : 0;
+
+            final Map<String, ?> sampleAtThetaPrime = takeSample(sampleFromVertices);
+
+            return new BuiltTree(
+                    leapfrog.position,
+                    leapfrog.gradient,
+                    leapfrog.momentum,
+                    logOfMasterPAfterLeapfrog,
+                    leapfrog.position,
+                    leapfrog.gradient,
+                    leapfrog.momentum,
+                    logOfMasterPAfterLeapfrog,
+                    leapfrog.position,
+                    leapfrog.gradient,
+                    logOfMasterPAfterLeapfrog,
+                    sampleAtThetaPrime,
+                    nPrime,
+                    sPrime
+            );
+
+        } else {
+            //Recursion—implicitly build the left and right subtrees.
+
+            BuiltTree leftTree = BuildTree(
+                    latentVertices,
+                    latentSetAndCascadeCache,
+                    probabilisticVertices,
+                    sampleFromVertices,
+                    position,
+                    gradient,
+                    momentum,
+                    u,
+                    v,
+                    j - 1,
+                    epsilon,
+                    random
+            );
+
+            Map<String, Double> positionBackward = leftTree.positionBackward;
```

Comment:
We've got an awful lot of maps cap'n.  As these maps are normally passed around and assigned in groups it gets a little hairy, e.g: 

`                    positionBackward = builtTree.positionBackward;`
`                    momentumBackward = builtTree.momentumBackward;`
`                    gradientBackward = builtTree.gradientBackward;`

You've got a `Leapfrogged` class that holds all of these maps, let's create an instance of this that holds position, momentum and gradient maps for a direction? 

Should make it simpler to pass things around and assign.

Then in your `BuiltTree` class, it would condense the 9 maps down to 3 maps and 2 `Leapfrogged` objects. We could call it something other than `Leapfrogged` if it gets confusing

Reply:
Turns out you can just reuse the map created down at the base case all the way up to the root. That cleans up all the maps and makes it easier to read.